DaemonSet e2e: Update image and rolling upgrade test timeout

alexbrand opened the pull request:
**What type of PR is this?**
/kind bug


**What this PR does / why we need it**: 
This PR removes an unnecessary sleep from the serve-hostname binary that is used in conformance tests.

The reason for this change is that the introduction of this 60 second sleep has negatively impacted the DaemonSet rolling upgrade conformance test. 

Specifically, it prevents the `serve-hostname` pods from terminating gracefully, and thus each pod takes at least 30 seconds to terminate (due to the default 30 second termination grace period). 

In large clusters, the conformance test fails because rolling all pods of the DaemonSet takes longer than timeout configured in the conformance test (5 minutes).

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #71666 

**Special notes for your reviewer**:
I have updated all the references to this image that I could find using code search. Not sure if I missed any.

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

ixdy commented on hack/testdata/pod-with-large-name.yaml:
```diff
@@ -8,4 +8,4 @@ metadata:
   containers:
   - name: kubernetes-serve-hostname
-    image: gcr.io/kubernetes-e2e-test-images/serve-hostname-amd64:1.1
+    image: gcr.io/kubernetes-e2e-test-images/serve-hostname-amd64:1.3
```
putting the reference updates into a separate commit might make this easier to cherrypick to release branches.

alexbrand replied:
Thanks @ixdy. Will move the updates to a separate commit.

alexbrand commented:
@freehan I think adding a flag is a possibility, but another potential route is just updating the DaemonSet test instead. I tagged you in the original issue as there is some more discussion happening there. Would love your insight and thoughts.

alexbrand committed:
DaemonSet e2e: Update image and rolling upgrade test timeout

Use Nginx as the DaemonSet image instead of the ServeHostname image.
This was changed because the ServeHostname has a sleep after terminating
which makes it incompatible with the DaemonSet Rolling Upgrade e2e test.

In addition, make the DaemonSet Rolling Upgrade e2e test timeout a
function of the number of nodes that make up the cluster. This is
required because the more nodes there are, the longer the time it will
take to complete a rolling upgrade.

Signed-off-by: Alexander Brand <alexbrand09@gmail.com>

alexbrand commented:
Updated the title to reflect the new approach for this fix. This is a more direct fix to the problem being observed in #71666. 

Instead of changing the `serve-hostname` image which is depended on by other tests, we fix the DaemonSet test directly. 

@BenTheElder, @ixdy, @freehan PTAL

ixdy commented:
/approve

I'll let @freehan apply lgtm.

janetkuo commented on test/e2e/apps/daemon_set.go:
```diff
@@ -329,6 +329,10 @@ var _ = SIGDescribe("Daemon set [Serial]", func() {
+		// Set the termination grace period to 1 second. This is necessary because the
+		// serve-hostname binary sleeps for 1 minute after receiving SIGTERM.
+		terminationPeriod := int64(1)
+		ds.Spec.Template.Spec.TerminationGracePeriodSeconds = &terminationPeriod
```
This fix should work, but DaemonSet test shouldn't be tied to the serve-hostname image. We can simply change the image we use in this test:
```
image := framework.ServeHostnameImage
```
to something else (can pick from [this list](https://github.com/kubernetes/kubernetes/blob/a2a5bd03fdae6ac45509560157f4c24d1c001e86/test/utils/image/manifest.go#L93)), such as nginx.

We can also increase the timeout based on number of nodes, as this affects the number of replicas a DaemonSet creates.

alexbrand replied:
That is a good point. Should we take that route instead? Will take a look at the list of images.

janetkuo replied:
Yes, thanks for looking into this!

alexbrand replied:
Should we update the image used throughout all DaemonSet tests, or just the one used during the DaemonSet update test?

janetkuo replied:
Updating the image used throughout all DaemonSet tests (i.e. change `image := framework.ServeHostnameImage`) is fine. Which image we pick doesn't matter. 

We can scale update timeout based on # of nodes too, maybe something like `5 mins + 1 min * # of nodes`.

alexbrand commented:
Updated the image, and made the timeout a function of the number of nodes in the cluster. PTAL.

janetkuo commented:
/lgtm

BenTheElder commented:
thank you!

janetkuo commented:
We need to cherrypick this to older releases as well.

janetkuo commented:
Sent cherrypick PRs: 
https://github.com/kubernetes/kubernetes/pull/74820
https://github.com/kubernetes/kubernetes/pull/74822
